src/Microsoft.ML.CpuMath/FactorizationMachine/AvxIntrinsics.cs

```diff
@@ -0,0 +1,188 @@
+                int j = pi[i];
+                linearResponse += pw[j] * px[i];
+
+                Vector256<float> x = Avx.BroadcastScalarToVector256(px + i);
+                Vector256<float> xx = Avx.Multiply(x, x);
+
+                // tmp -= <v_j,f, v_j,f> * x * x
+                int vBias = j * m * d + f * d;
+
+                // j-th feature's latent vector in the f-th field hidden space.
+                float* vjf = pv + vBias;
+
+                for (int k = 0; k + 8 <= d; k += 8)
+                {
+                    Vector256<float> vjfBuffer = Avx.LoadVector256(vjf + k);
+                    tmp = Avx.Subtract(tmp, Avx.Multiply(Avx.Multiply(vjfBuffer, vjfBuffer), xx));
+                }
+
+                for (int fprime = 0; fprime < m; fprime++)
+                {
+                    vBias = j * m * d + fprime * d;
+                    int qBias = f * m * d + fprime * d;
+                    float* vjfprime = pv + vBias;
+                    float* qffprime = pq + qBias;
+
+                    // q_f,f' += v_j,f' * x
+                    for (int k = 0; k + 8 <= d; k += 8)
+                    {
+                        Vector256<float> vjfprimeBuffer = Avx.LoadVector256(vjfprime + k);
+                        Vector256<float> q = Avx.LoadVector256(qffprime + k);
+                        q = Avx.Add(q, Avx.Multiply(vjfprimeBuffer, x));
+                        Avx.Store(qffprime + k, q);
+                    }
+                }
+            }
+
+            for (int f = 0; f < m; f++)
+            {
+                // tmp += <q_f,f, q_f,f>
+                float* qff = pq + f * m * d + f * d;
+                for (int k = 0; k + 8 <= d; k += 8)
+                {
+                    Vector256<float> qffBuffer = Avx.LoadVector256(qff + k);
+
+                    // Intra-field interactions.
+                    tmp = Avx.Add(tmp, Avx.Multiply(qffBuffer, qffBuffer));
+                }
+
+                // y += <q_f,f', q_f',f>, f != f'
+                // Whis loop handles inter - field interactions because f != f'.
+                for (int fprime = f + 1; fprime < m; fprime++)
+                {
+                    float* qffprime = pq + f * m * d + fprime * d;
+                    float* qfprimef = pq + fprime * m * d + f * d;
+                    for (int k = 0; k + 8 <= d; k += 8)
+                    {
+                        // Inter-field interaction.
+                        Vector256<float> qffprimeBuffer = Avx.LoadVector256(qffprime + k);
+                        Vector256<float> qfprimefBuffer = Avx.LoadVector256(qfprimef + k);
+                        y = Avx.Add(y, Avx.Multiply(qffprimeBuffer, qfprimefBuffer));
+                    }
+                }
+            }
+
+            y = Avx.Add(y, Avx.Multiply(Vector256.Create(0.5f), tmp));
+            tmp = Avx.Add(y, Avx.Permute2x128(y, y, 1));
+            tmp = Avx.HorizontalAdd(tmp, tmp);
+            y = Avx.HorizontalAdd(tmp, tmp);
```

Comment:
Why the difference here from the C++ code?

In C++ we have:

```C++
_tmp = _mm_add_ps(_y, _mm_movehl_ps(_y, _y));
_y = _mm_add_ps(_tmp, _mm_shuffle_ps(_tmp, _tmp, 1)); // The lowest slot is the response value.
```

But here, we have:

```C#
tmp = Avx.Add(y, Avx.Permute2x128(y, y, 1));
tmp = Avx.HorizontalAdd(tmp, tmp);
y = Avx.HorizontalAdd(tmp, tmp);
```

Every other line in this method reads line-for-line identical to the C++ code (besides the SSE vs AVX differences).

Reply:
There is no AVX equivalent instruction for `_mm_movehl_ps(_y, _y)`.  Permuting and doing horizontal add on the two floating point elements (`tmp`) does the same computation.